### PR TITLE
jobmanager(engine):support WatchJobStatuses

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: ref #5513

### What is changed and how it works?
- Implement `WatchJobStatuses` for `JobManagerImplV2`.

### Review's Guide
- This PR aims to support local resource GC by enabling the `JobManager` to report all events of jobs being removed.
- The `JobManager` needs to be further refactored to make it more robust. This PR only aims to implement `WatchJobStatuses`, but not to optimize concurrency control or latency.

#### Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/servermaster/jobmanager.go [hunk 1]
```diff
@@ -431,5 +467,36 @@ func (jm *JobManagerImplV2) OnWorkerStatusUpdated(worker lib.WorkerHandle, newSt
...
 }
+
+// WatchJobStatuses returns a snapshot of job statuses followed by a stream
+// of job status changes.
+func (jm *JobManagerImplV2) WatchJobStatuses(
+	ctx context.Context,
+) (resManager.JobStatusesSnapshot, *notifier.Receiver[resManager.JobStatusChangeEvent], error) {
+	// We add an explicit deadline to make sure that
+	// any potential problem will not block the JobManager forever.
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
```
Comment (maxshuang) [inline engine/servermaster/jobmanager.go:481]: When `SubmitJob` is called, 
(1) Should we call `WatchJobStatuses` again? Or just add to the job snapshot?
(2) Should we notify the notifier?

File: engine/servermaster/jobmanager.go [hunk 2]
```diff
@@ -118,19 +133,38 @@ func (jm *JobManagerImplV2) CancelJob(ctx context.Context, req *pb.CancelJobRequ
...
 	}
+
+	return &pb.CancelJobResponse{}
+}
+
+func (jm *JobManagerImplV2) deleteJobMeta(ctx context.Context, jobID string) error {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	jm.jobStatusChangeMu.Lock(ctx)
```
Comment (gozssky) [inline engine/servermaster/jobmanager.go:150]: Should check the return value?

File: engine/servermaster/jobmanager.go [hunk 3]
```diff
@@ -431,5 +467,36 @@ func (jm *JobManagerImplV2) OnWorkerStatusUpdated(worker lib.WorkerHandle, newSt
...
+func (jm *JobManagerImplV2) WatchJobStatuses(
+	ctx context.Context,
+) (resManager.JobStatusesSnapshot, *notifier.Receiver[resManager.JobStatusChangeEvent], error) {
+	// We add an explicit deadline to make sure that
+	// any potential problem will not block the JobManager forever.
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	// Note that the lock is cancellable by the context.
+	jm.jobStatusChangeMu.Lock(ctx)
```
Comment (gozssky) [inline engine/servermaster/jobmanager.go:485]: ditto

File: engine/servermaster/jobmanager.go (hunk 1 above)
Comment (liuzix) [reply engine/servermaster/jobmanager.go:484]: - We will postpone implementing notification of the `SubmitJob` event.
- The caller of `WatchJobStatuses` is told about any changes by the `receiver`. JobManager only needs the add a notification to the `notifier`.

**Review by gozssky — APPROVED** [review]

**Review by maxshuang — APPROVED** [review]

**Pull request merged**